File: ArimaaIA02/apps/arimaa/arimaa.pl
:- module(bot,
      [  get_moves/3
      ]).

% Parsing
% - dos2unix to the file
% - replace (A \= B) par \+ (A == B)
% - for negation always use \+(Goal)
% - for all operation or comparaison (+, -, > ...) use parenthesis: X is (Y-1)
% - No native like random or nth0
% - no _ in move_piece([],[],_,_).
% - Never have possibility to ?



% board: [[0,0,rabbit,silver],[0,1,rabbit,silver],[0,2,horse,silver],[0,3,rabbit,silver],[0,4,elephant,silver],[0,5,rabbit,silver],[0,6,rabbit,silver],[0,7,rabbit,silver],[1,0,camel,silver],[1,1,cat,silver],[1,2,rabbit,silver],[1,3,dog,silver],[1,4,rabbit,silver],[1,5,horse,silver],[1,6,dog,silver],[1,7,cat,silver],[2,7,rabbit,gold],[6,0,cat,gold],[6,1,horse,gold],[6,2,camel,gold],[6,3,elephant,gold],[6,4,rabbit,gold],[6,5,dog,gold],[6,6,rabbit,gold],[7,0,rabbit,gold],[7,1,rabbit,gold],[7,2,rabbit,gold],[7,3,cat,gold],[7,4,dog,gold],[7,5,rabbit,gold],[7,6,horse,gold],[7,7,rabbit,gold]]
% gamestate: [side, [captured pieces] ] (e.g. [silver, [ [0,1,rabbit,silver],[0,2,horse,silver] ]) 

% get_moves(Moves, [silver, []], [[0,0,rabbit,silver],[0,1,rabbit,silver],[6,0,cat,gold],[6,1,horse,gold]]).

%get_moves([[0,0]],Gamestate,Board) :- generate_moves(Gamestate,Board,Moves,4).
get_moves(Moves,Gamestate,Board) :- generate_moves(Gamestate,Board,Moves,4).

% get_moves(Moves, [silver, []], [[0,0,rabbit,silver],[0,1,rabbit,silver],[0,2,horse,silver],[0,3,rabbit,silver],[0,4,elephant,silver],[0,5,rabbit,silver],[0,6,rabbit,silver],[0,7,rabbit,silver],[1,0,camel,silver],[1,1,cat,silver],[1,2,rabbit,silver],[1,3,dog,silver],[1,4,rabbit,silver],[1,5,horse,silver],[1,6,dog,silver],[1,7,cat,silver],[2,7,rabbit,gold],[6,0,cat,gold],[6,1,horse,gold],[6,2,camel,gold],[6,3,elephant,gold],[6,4,rabbit,gold],[6,5,dog,gold],[6,6,rabbit,gold],[7,0,rabbit,gold],[7,1,rabbit,gold],[7,2,rabbit,gold],[7,3,cat,gold],[7,4,dog,gold],[7,5,rabbit,gold],[7,6,horse,gold],[7,7,rabbit,gold]]).

% generate_moves([silver, []], [[0,0,rabbit,silver],[0,1,rabbit,silver],[6,0,cat,gold],[6,1,horse,gold]], Moves, 2).

% comp([[0,0,rabbit,silver],[0,1,rabbit,silver],[0,2,horse,silver],[0,3,rabbit,silver],[0,4,elephant,silver],[0,5,rabbit,silver],[0,6,rabbit,silver],[0,7,rabbit,silver],[1,0,camel,silver],[1,1,cat,silver],[1,2,rabbit,silver],[1,3,dog,silver],[1,4,rabbit,silver],[1,5,horse,silver],[1,6,dog,silver],[1,7,cat,silver],[2,7,rabbit,gold],[6,0,cat,gold],[6,1,horse,gold],[6,2,camel,gold],[6,3,elephant,gold],[6,4,rabbit,gold],[6,5,dog,gold],[6,6,rabbit,gold]], [[0,0,rabbit,silver],[0,1,rabbit,silver],[0,2,horse,silver],[0,3,rabbit,silver],[0,4,elephant,silver],[0,5,rabbit,silver],[0,6,rabbit,silver],[0,7,rabbit,silver],[1,0,camel,silver],[1,1,cat,silver],[1,2,rabbit,silver],[1,3,dog,silver],[1,4,rabbit,silver],[1,5,horse,silver],[1,6,dog,silver],[1,7,cat,silver],[2,7,rabbit,gold],[6,0,cat,gold],[6,1,horse,gold],[6,2,camel,gold],[6,3,elephant,gold],[6,4,rabbit,gold],[6,5,dog,gold],[6,6,rabbit,gold]], X).

generate_moves(_,_,[],0).
generate_moves(Gamestate,Board,[[From,To]|Moves],N) :- M is (N-1),
											get_a_random_ally_piece(Gamestate,Board,From),
											is_legal_move(Gamestate,Board,From,To),
											move_piece(Board,NewBoard,From,To),
											comp(Board,NewBoard,Diff), Diff \= [], % at least a significant moves has been done (has to be compared with the initial board)
											generate_moves(Gamestate,NewBoard,Moves,M), !.

:- dynamic(previous_moves/1).
  
previous_moves([]).
add_to_previous_moves(Move) :- previous_moves(X), retractall(previous_moves(_)), asserta(previous_moves([Move|X])).
											
strength(rabbit,1).
strength(cat,2).
strength(dog,3).
strength(horse,4).
strength(camel,5).
strength(elephant,6).
gold_home_row(5).
silver_home_row(0).
		
member(X, [X|_]).
member(X, [_|Tail]) :-  member(X, Tail). 

comp([], _, []) :- !.
comp(L, [], L) :- !.
comp([H|T], S, R) :-
    (   member(H, S)
    ->  comp(T, S, R)
    ;   R = [H|RT],
        comp(T, S, RT)
    ).
	
inter([], _, []) :- !.
inter([H1|T1], L2, [H1|Res]) :- member(H1, L2), !, inter(T1, L2, Res).
inter([_|T1], L2, Res) :- inter(T1, L2, Res).
  
findx(0,[Elem|_],Elem):- !.
findx(N,[_|List],Elem) :- findx(N1,List,Elem), N is (N1+1).
		
len([], 0) :- !.
len([_|L], Length) :- len(L, N1), Length is (N1+1).
		
choose_random([], []).
choose_random(List, Elt) :-
        len(List, Length),
        random(0, Length, Index),
        findx(Index, List, Elt).
		
choose_piece([Elt|_], Elt).
choose_piece([_|List], Elt) :- choose_piece(List, Elt).

% For all board piece, if the piece is tagged with MySide put it in the resulting list
get_all_ally_piece(_,[],[]).
get_all_ally_piece([MySide|Tail],[[Row,Col,_,MySide]|Board],[[Row,Col]|Pieces]) :- get_all_ally_piece([MySide|Tail],Board,Pieces).
get_all_ally_piece(Gamestate,[_|Board],Pieces) :- get_all_ally_piece(Gamestate,Board,Pieces).

get_a_random_ally_piece(Gamestate,Board,Piece) :- get_all_ally_piece(Gamestate,Board,Pieces), choose_random(Pieces,Piece).

% move_piece(Board,NewBoard,From,To)
move_piece([],[],_,_).
move_piece([[Row,Col,Type,Side]|Board],[[RowTo,ColTo,Type,Side]|NewBoard],[Row,Col],[RowTo,ColTo]) :- move_piece(Board, NewBoard, [Row,Col],[RowTo,ColTo]).
move_piece([[Row,Col,Type,Side]|Board],[[Row,Col,Type,Side]|NewBoard],[RowFrom,ColFrom],[RowTo,ColTo]) :- move_piece(Board, NewBoard, [RowFrom,ColFrom],[RowTo,ColTo]).

% Return false if piece does not exist
% get_piece_and_side_from_coord(Board,Coord,Piece,Side)
%get_piece_and_side_from_coord([],Coord,Type,Side).
get_piece_and_side_from_coord([[Row,Col,Type,Side]|_],[Row,Col],Type,Side).
get_piece_and_side_from_coord([_|Tail],Coord,Type,Side) :- get_piece_and_side_from_coord(Tail,Coord,Type,Side).

% get_neighbour_coord(Board, FromCoord, ToCoord)
get_neighbour_coord(Board, [R,C], [Rto,C]) :- \+(get_piece_and_side_from_coord(Board,[R,C],rabbit,gold)), (R < 7), Rto is (R+1).
get_neighbour_coord(_, [R,C], [R,Cto]) :- (C < 7), Cto is (C+1).
get_neighbour_coord(Board, [R,C], [Rto,C]) :- \+(get_piece_and_side_from_coord(Board,[R,C],rabbit,silver)), (R > 0), Rto is (R-1).
get_neighbour_coord(_, [R,C], [R,Cto]) :- (C > 0), Cto is (C-1).

% has_friendly_neighbour(Board, Coord)
has_friendly_neighbour(Board, Coord) :- get_neighbour_coord(Board, Coord, NeighbourCoord), get_piece_and_side_from_coord(Board, Coord, _, Side), get_piece_and_side_from_coord(Board, NeighbourCoord, _, Side).

has_enemie_neighbour_stronger(Board, Coord) :- get_neighbour_coord(Board, Coord, NeighbourCoord), 
											   get_piece_and_side_from_coord(Board, Coord, MyType, Side), 
										       get_piece_and_side_from_coord(Board, NeighbourCoord, EnemieType, EnemieSide),	
											   Side \= EnemieSide, strength(EnemieType,X1), strength(MyType,X2), X1 > X2.
											  
% is_frozen(Board, Coord)
is_frozen(Board, Coord) :- has_friendly_neighbour(Board, Coord), !, fail.
is_frozen(Board, Coord) :- has_enemie_neighbour_stronger(Board, Coord).
										  

% if silver_rabbit goes to gold home row or gold rabbit goes to silver home row
% is_gameover(Gamestate,Board)
is_gameover([MySide|_],[[Row,_,Type,Side]|_]) :- MySide == gold, gold_home_row(Row), Side == silver, Type == rabbit.
is_gameover([MySide|_],[[Row,_,Type,Side]|_]) :- MySide == silver, silver_home_row(Row), Side == gold, Type == rabbit.
is_gameover(Gamestate,[_|Tail]) :- is_gameover(Gamestate,Tail).

is_victory([MySide|_],[[Row,_,Type,Side]|_]) :- MySide == silver, gold_home_row(Row), Side == silver, Type == rabbit.
is_victory([MySide|_],[[Row,_,Type,Side]|_]) :- MySide == gold, silver_home_row(Row), Side == gold, Type == rabbit.
is_victory(Gamestate,[_|Tail]) :- is_victory(Gamestate,Tail).

% get_legal_move(Gamestate,Board,FromCoord,Moves)
% If game over, return empty
%get_legal_move(Gamestate,Board,_,[]) :- is_gameover(Gamestate,Board).
% if from coord do not contains piece, return empty
%get_legal_move(Gamestate,Board,FromCoord,[]) :- \+ get_piece_and_side_from_coord(Board,FromCoord,_,_).

is_case_busy([[Row,Col,_,_]|_],[Row,Col]). 
is_case_busy([_|Board],Piece) :- is_case_busy(Board,Piece).

% is_legal_move(Gamestate,Board,FromCoord,Move)
% If game over, return empty
is_legal_move([_|_],Board,From,To) :- 
						get_neighbour_coord(Board,From,To), % The to is a neighbour
						\+(is_case_busy(Board,To)). % check the to case is available , TODO: or it is weaker enemie piece that I can take place
						% previous_moves(Moves), \+(findx(_,Moves,To)). % Check if it is not one of my previous move

File: ArimaaIA02/apps/arimaa/js/arimaa.js
var ARIMAA = ARIMAA || function() {

	var board_width = 8;
	var board_height = 8;
  var	silver = { 'side': 'silver' };
  var gold = { 'side': 'gold' };
  
  var silver_homerow = 0;
  var gold_homerow = board_height - 1;
  
  var rabbit = create_piece('rabbit', 1);
  var cat = create_piece('cat', 2);
  var dog = create_piece('dog', 3);
  var horse = create_piece('horse', 4);
  var camel = create_piece('camel', 5);
  var elephant = create_piece('elephant', 6);
  
  var steps_in_move = 4;
  var steps_in_setting_move = 16;

  var traps = [[2, 2], [2, 5], [5, 2], [5, 5]];
  
	var west = [-1, 0];
	var east = [1, 0];
	var north = [0, -1];
	var south = [0, 1];
	
	// maybe y should be in 0 for consistency (wrt. board)
	// also for traps, though those are symmetric atm
	function get_x(direction) { return direction[0]; }
	function get_y(direction) { return direction[1]; }
	
	var directions = [west, east, north, south];
	
	function trap_as_coordinate(trap) {
		return {
			'col': trap[0],
			'row': trap[1]
		}
	}
	function is_rabbit(piece) { return piece.type === 'rabbit'; }
  
  function create_piece(type, strength) {
  	return {
  		'type': type,
  		'strength': strength
  	}
  }
  
  /** Copies the board structure (immutability guarantees less problems) */
  function copy_board(old) {
  	return GENERIC.map(old, function(row) {
  		return GENERIC.map(row, function(elem) { return elem; });
  	});
  }
  
  function is_in_trap(coordinate) {
  	return GENERIC.exists(traps, function(trap) {
  			return coordinate['row'] === trap[0] && coordinate['col'] === trap[1];
  	});
  }
  
  function is_coordinate_capturing(coordinate, board) {
  	var piece = get_piece(coordinate, board);
  	var exists_friendly_neighbour = GENERIC.exists(neighbours(coordinate, board), function(neighbour) {
  			return is_friendly(piece, neighbour);
    });

		return !exists_friendly_neighbour;  
  }

	function do_capturing(board) {
		var captured = [];
		
		GENERIC.for_each(traps, function(trap) {
			var trap_coord = trap_as_coordinate(trap);
			var piece = get_piece(trap_coord, board);
				 
			if(!!piece) {
				if(is_coordinate_capturing(trap_coord, board)) {
					captured.push(board[trap_coord.row][trap_coord.col]);
					board[trap_coord.row][trap_coord.col] = {} // captures piece
				}
			 }
		 });
		
		return captured;
	}

  function move_piece(gamestate, board, piece_coordinate, new_coordinate) {
  	 var new_board = copy_board(board);

  	 if(piece_coordinate.row === undefined || new_board[piece_coordinate.row] === undefined){
  	 	 //GENERIC.log("row", piece_coordinate.row);
  	 	 throw "illegal or undefined piece coordinate";
  	 }
  	 
  	 var piece = new_board[piece_coordinate.row][piece_coordinate.col];
  	 new_board[piece_coordinate.row][piece_coordinate.col] = {}; // takes piece away from old place
  	 if(!piece) throw "moved empty piece";
  	 new_board[new_coordinate.row][new_coordinate.col] = piece;
  	 
		 var captured = do_capturing(new_board);
  	 
  	 var laststep = {
  	 	 'piece': piece,
  	 	 'from': piece_coordinate,
  	 	 'to': new_coordinate  	 	 
  	 }
  	 
  	 return {
  	 	 'board': new_board,
  	 	 'gamestate': gamestate_after_move(gamestate, laststep, captured)
  	 }
  }
  
  function same_coordinates(a, b) {
  	return a.row === b.row && a.col === b.col;
  }

  function neighbours(coordinate, board) {
  	var neighbour_candidates = GENERIC.map(neighbour_coordinates(coordinate), function(neighbour_coordinate) {
  			return get_piece(neighbour_coordinate, board);
  	});

  	return GENERIC.filter(neighbour_candidates, function(candidate) { return !!candidate; });
  }
  
  function neighbour_coordinates(coordinate) {
  	return GENERIC.filter(
  					GENERIC.map(directions, function(direction) { return get_neighbour_coordinate(coordinate, direction); }),
  					function(elem) { return !!elem; });
  }
  
  function gamestate_after_move(gamestate, laststep, captured) {
  	var copy = GENERIC.shallowCopyObject(gamestate);
  	
  	if(!!gamestate.expectedmove) copy.expectedmove = undefined; // clear expectations for thise turn
  		
  	if(!gamestate.captured) {
  		copy.captured = captured;
  	} else {
  		copy.captured = copy.captured.concat(captured);
  	}
  	
  	if(gamestate.steps > 1) {
  		copy.steps = gamestate.steps - 1;
    	copy.laststep = laststep;
    	//FIXME: this is too complicated to be here
    	// if it's the case of pushing, we must set expectations for next turn to it be completed
    	if(laststep !== undefined && laststep.piece.side !== gamestate.turn && 
    		(gamestate.laststep === undefined || !same_coordinates(gamestate.laststep.from, laststep.to))) { // the case where pulling was completed must be checked
    		copy.expectedmove = {
    			'side': gamestate.turn,
    			'strength': laststep.piece.strength,
    			'from_array': neighbour_coordinates(laststep.from),
    			'to': laststep.from
    		}     		
    	}
  	} else {
  	  copy.steps = steps_in_move;
  	  copy.turn = gamestate.turn === gold ? silver : gold;
  	  copy.laststep = undefined;
  	}  		
  	
 		return copy;
  }
  
  function is_empty_square(square) { return square.type === undefined; }
  
  function is_stronger(piece1, piece2) {
  	return piece1.strength > piece2.strength;
  }
  
  function is_friendly(piece1, piece2) { return piece1.side === piece2.side; }
  
  function get_piece(coordinate, board) { 
  	var square = board[coordinate.row][coordinate.col];
  	return is_empty_square(square) ? false : square; 
  }
  
  function is_in_board(coordinate) { return coordinate.row >= 0 && coordinate.row < board_height && coordinate.col >= 0 && coordinate.col < board_width; }

  function get_neighbour_coordinate(coordinate, direction) {
		var neighbour_coordinate = { 
			'col': coordinate.col + get_x(direction),
			'row': coordinate.row + get_y(direction)
		}
		
		if(!is_in_board(neighbour_coordinate)) return false;
		else return neighbour_coordinate;
  }
  
  function get_neighbour(coordinate, direction, board) {
  	var coord = get_neighbour_coordinate(coordinate, direction);
  	if(!coord) return false;
		else return get_piece(coord, board);
  }
  
  function is_frozen(coordinate, board) {
    var piece = get_piece(coordinate, board);
    
    var friendly_neighbour_exists = GENERIC.exists(directions, function(direction) {
    		var neighbour = get_neighbour(coordinate, direction, board);
    		return !!neighbour && is_friendly(piece, neighbour);
    });
    
    if(friendly_neighbour_exists) return false;
    // check if there's opponent neighbour that is stronger
    else return GENERIC.exists(directions, function(direction) {
    	var neighbour = get_neighbour(coordinate, direction, board);
    	return !!neighbour && is_stronger(neighbour, piece) && !is_friendly(piece, neighbour);
    });
  }

  function is_gold_rabbit_at_goal(board) {
  	return GENERIC.exists(board[silver_homerow], function(elem) {
  		return is_rabbit(elem) && elem.side === ARIMAA.gold;
  	});
  }
  
  function is_gameover(board, gamestate) {
  	// FIXME: check repetetion rule, maybe also resigning from gamestate?
  	var silver_rabbit_in_goal = GENERIC.exists(board[gold_homerow], function(elem) {
  		return is_rabbit(elem) && elem.side === ARIMAA.silver;
  	});
  	
  	var gold_rabbit_in_goal = GENERIC.exists(board[silver_homerow], function(elem) {
  		return is_rabbit(elem) && elem.side === ARIMAA.gold;
  	});
  	
  	// the realisation of loss is only after turn change
  	var rabbit_in_goal = gamestate.turn === gold && silver_rabbit_in_goal 
  											|| gamestate.turn === silver && gold_rabbit_in_goal;
  	
  	return rabbit_in_goal;
  }
  
  function current_player_piece(coordinate, board, gamestate) {
  	return get_piece(coordinate, board).side === gamestate.turn;
  }
  
  // assumption: piece in coordinate is opponents and steps >= 2
  // in push moves, opponent is moves first always
  function push_moves(gamestate, board, coordinate) {
  	// there is an empty square next to this coordinate where a neighbour opponent can push or pull me
  	var neighbour_coords = neighbour_coordinates(coordinate);
  	
  	var empty_squares = GENERIC.filter(neighbour_coords, function(neighbour_coordinate) {
  			return is_empty_square(get_piece(neighbour_coordinate, board));
  	});
  	
  	if(empty_squares.length === 0) return [];

  	var piece = get_piece(coordinate, board);
  	
  	var exists_stronger_neighbour = GENERIC.exists(neighbour_coords, function(neighbour_coord) {
  			var neighbour = get_piece(neighbour_coord, board);
  			return !is_friendly(piece, neighbour) && is_stronger(neighbour, piece) && !is_frozen(neighbour_coord, board);
  	});  	
  	
  	return exists_stronger_neighbour ? empty_squares : [];
  }
  
  function pull_moves(gamestate, board, coordinate) {
  	if(!gamestate.laststep) return []; // no last step in this side
  	if(!is_stronger(gamestate.laststep.piece, get_piece(coordinate, board))) return [];

    // check whether last move there was a move by current side that can be considered pull in this step
  	var can_be_pulled = GENERIC.exists(directions, function(direction) {
  			var neighbour_coord = get_neighbour_coordinate(coordinate, direction);
  			if(neighbour_coord.col === gamestate.laststep.from.col && neighbour_coord.row === gamestate.laststep.from.row) return true;
  	});
  	
  	if(can_be_pulled) {
  		return [gamestate.laststep.from];
  	} else return []; 	
  }
  
  function is_legal_rabbit_move(direction, gamestate) {
  	return (gamestate.turn === gold && get_y(direction) <= 0) || (gamestate.turn === silver && get_y(direction) >= 0);
  }

  function is_passing_legal(gamestate, board) {
  	return !is_gameover(board, gamestate)
  					&& gamestate.type !== 'setting'
  					&& gamestate.steps < steps_in_move // at least one move is obligatory
  					&& gamestate.expectedmove === undefined; // if opponent's piece was pushed, the pusher must be moved
  }
  
  function legal_moves(gamestate, board, coordinate) {
  	if(is_gameover(board, gamestate)) return [];
  	if(board[coordinate.row][coordinate.col].type === undefined) return [];

  	// expected move means that one must complete push
  	if(!!gamestate.expectedmove) {
  		if(gamestate.turn !== gamestate.expectedmove.side) return [];
  	  if(get_piece(coordinate, board).strength <= gamestate.expectedmove.strength) return [];
   	  if(get_piece(coordinate, board).side !== gamestate.turn) return [];
  	  var is_from_neighbour = GENERIC.exists(gamestate.expectedmove.from_array, function(from) {
  	  		return same_coordinates(from, coordinate);
  	  });
  	  if(is_from_neighbour) return [gamestate.expectedmove.to];
  	  else return [];
    }
    
    var result = [];
    
  	var is_piece_current_players = current_player_piece(coordinate, board, gamestate);
  	if(is_piece_current_players && is_frozen(coordinate, board)) return [];
  	
  	if(!is_piece_current_players) {
     	var pull_result = pull_moves(gamestate, board, coordinate);
  		
  		if(gamestate.steps >= 2) {
  		  return pull_result.concat( push_moves(gamestate, board, coordinate) );
  		} else return pull_result;
  	}
  	
  	function can_move_to(direction) {
  	  if(is_rabbit(get_piece(coordinate, board)) && !is_legal_rabbit_move(direction, gamestate)) { return false; }
  
  		var x = coordinate.col + get_x(direction);
  		var y = coordinate.row + get_y(direction);
  		
  		if(x < 0 || y < 0 || x >= board_width || y >= board_height) return false;
  		return is_empty_square(board[y][x]); //FIXME: also pushing and pulling should be considered
  	}
  	
  	var legal_directions = GENERIC.filter([west, east, north, south], can_move_to);
  	return GENERIC.map(legal_directions, function(direction) { 
  			return {
  				'row': coordinate.row + get_y(direction),
  				'col': coordinate.col + get_x(direction)
  			}
  	});
  }
  
  function get_new_coordinate(from, direction) {
  	return {
  		'col': from.col + get_x(direction),
  		'row': from.row + get_y(direction)
  	}
  }
  
  function get_piece_with_side(piece, side) {
  	var copy = GENERIC.shallowCopyObject(piece);
  	copy['side'] = side;
  	return copy;
  }
  
  function add_piece(piece, coordinate, board, gamestate) {
  	var new_board = copy_board(board);
  	var new_gamestate = GENERIC.shallowCopyObject(gamestate);

  	new_board[coordinate.row][coordinate.col] = piece;

  	new_gamestate.steps--;

		if(new_gamestate.steps < 0) {
			GENERIC.log(new_gamestate);
			throw "steps < 0";
		}
		
  	if(new_gamestate.steps === 0) {
  		if(gamestate.turn === gold) {
  			new_gamestate.steps = steps_in_setting_move;
  		} else {
  			new_gamestate.steps = steps_in_move;
  			new_gamestate.type = 'normal';
  		}

  	  new_gamestate.turn = gamestate.turn === gold ? silver : gold;
  	}
  	
  	return {
  		'board': new_board,
  		'gamestate': new_gamestate
  	}
  }
  
  function remove_piece(coordinate, board) {
  	var new_board = copy_board(board);
  	new_board[coordinate.row][coordinate.col] = {};
  	return new_board;
  }
  
  function get_initial_gamestate() {
  	return {
	  'turn': gold,
	  'steps': steps_in_setting_move,
	  'type': 'setting'
	  }
	}
	
	function pass(board, gamestate) {
		var new_gamestate = GENERIC.shallowCopyObject(gamestate);
		new_gamestate.turn = gamestate.turn === gold ? silver : gold;
		new_gamestate.steps = steps_in_move;
		new_gamestate.laststep = undefined;
		
		return {
			'board': board,
			'gamestate': new_gamestate
		}
	}
  
	function opposite_turn(turn) { return turn === ARIMAA.gold ? ARIMAA.silver : ARIMAA.gold }

  return {
  	'board_width': board_width,
  	'board_height': board_height,
  	'steps_in_move': steps_in_move,
  	'steps_in_setting_move': steps_in_setting_move,
  	'traps': traps,
  	'directions': { 'east': east, 'west': west, 'north': north, 'south': south},
  	'get_new_coordinate': get_new_coordinate,
  	'silver': silver,
  	'gold': gold,
  	'rabbit': rabbit,
  	'cat': cat,
  	'dog': dog,
  	'horse': horse,
  	'camel': camel,
  	'elephant': elephant,
  	'legal_moves': legal_moves,
  	'move_piece': move_piece,
  	'get_piece_with_side': get_piece_with_side,
  	'add_piece': add_piece,
  	'remove_piece': remove_piece,
  	'get_initial_gamestate': get_initial_gamestate,
  	'pass': pass,
  	'is_passing_legal': is_passing_legal,
  	'opposite_turn': opposite_turn,
  	'is_gameover': is_gameover,
  	'is_gold_rabbit_at_goal': is_gold_rabbit_at_goal
  }
  
}();